src/main/java/org/radarbase/kafka/connect/transforms/KeyValueTransform.java

```diff
@@ -0,0 +1,125 @@
+  private static final String PURPOSE = "copying fields from value to key";
+  private static final String TIMESTAMP_FIELD = "timestamp";
+  private static final String KEYVALUE_SCHEMA_NAME = "KeyToValue";
+  private static final String TIME_FIELD = "time";
+
+  @Override
+  public R apply(R r) {
+    if (r.valueSchema() == null) {
+      return applySchemaless(r);
+    } else {
+      return applyWithSchema(r);
+    }
+  }
+
+  private R applyWithSchema(R r) {
+    SchemaBuilder schemaBuilder = SchemaBuilder.struct().name(KEYVALUE_SCHEMA_NAME);
+    schemaBuilder = updateSchema(schemaBuilder, r.keySchema());
+    schemaBuilder = updateSchema(schemaBuilder, r.valueSchema());
+    schemaBuilder.field(TIMESTAMP_FIELD, Schema.INT64_SCHEMA);
+    Schema schema = schemaBuilder.build();
+
+    final Struct recordValue = requireStruct(r.value(), PURPOSE);
+    final Struct recordKey = requireStruct(r.key(), PURPOSE);
+    Struct newData = new Struct(schema);
+    newData = addValuesToNewSchema(newData, recordKey, r.keySchema());
+    newData = addValuesToNewSchema(newData, recordValue, r.valueSchema());
+    newData.put(TIMESTAMP_FIELD, r.timestamp());
+
+    return r.newRecord(r.topic(), r.kafkaPartition(), null, null, schema, newData, r.timestamp());
+  }
+
+  private Struct addValuesToNewSchema(Struct newData, Struct oldData, Schema oldSchema){
+    for (Field field: oldSchema.fields()) {
+      String fieldName = field.name();
+      Object fieldVal = oldData.get(field);
+      if(fieldName.contains(TIME_FIELD))
+        fieldVal = convertTimestamp(fieldVal);
+      newData.put(fieldName, fieldVal);
+    }
+    return newData;
+  }
+
+
+  private SchemaBuilder updateSchema(SchemaBuilder schemaBuilder, Schema oldSchema){
+    for (Field field: oldSchema.fields()) {
+      String fieldName = field.name();
+      if(fieldName.contains(TIME_FIELD))
+        schemaBuilder.field(fieldName, Schema.INT64_SCHEMA);
+      else
+        schemaBuilder.field(fieldName, field.schema());
+    }
+    return schemaBuilder;
+  }
+
+  private R applySchemaless(R record) {
+    Map<String, Object> value = requireMap(record.value(), PURPOSE);
+    Map<String, Object> key = requireMap(record.key(), PURPOSE);
+    Map<String, Object> newData = new HashMap<>();
+    for (Map.Entry<String, Object> entry : key.entrySet()) {
+      newData.put(entry.getKey(), entry.getValue());
+    }
+    for (Map.Entry<String, Object> entry : value.entrySet()) {
+      String fieldName = entry.getKey();
+      Object fieldVal = entry.getValue();
+      if(fieldName.contains(TIME_FIELD))
+        fieldVal = convertTimestamp(fieldVal);
+      newData.put(fieldName, fieldVal);
+    }
+    newData.put(TIMESTAMP_FIELD, record.timestamp());
+    return record.newRecord(record.topic(), record.kafkaPartition(), null, null, null, newData, record.timestamp());
+  }
+
+  private long convertTimestamp(Object time){
+    return (long) (Double.parseDouble(time.toString()) * 1000);
```

Comment:
Perhaps use a branch here? If it is a double, no need to parse double. If it’s not, numberformatexception should be handled.

Reply:
Ok I've change it now, but I'm not sure if [this](https://github.com/mpgxvii/kafka-connect-transform/pull/1/commits/4de443090717979dc3ca8234db165346ffaa7fe8#diff-5e08fff00fb5d36593c4e921f5790342R107) is what you meant?